---
title: "R Notebook"
output: html_notebook
---

## Introduction

The analysis was conducted on 20 Italian samples (4 from carcass, 4 from clean drain, 4 from clean surface, 
4 from dirty drain, and 4 from dirty surface). For these 20 samples, we have the following information available:

* The abundances (as counts) and the taxonomy (via *taxizedb*) of the bacteria obtained from read files.
* The abundances (as counts) of the ARG genes derived from read files. Here, we have only the information for the different arg genes.
* The abundances (as coverages) of the ARG genes obtained from contigs. Here, we have additional details, including:
    + ARG gene classification.
    + Contig coverage, used as the value of ARG abundance.
    + Bacterial taxonomy of the contigs containing the ARG genes (*Diamond_nr*).
    + Waafle plasmid: a logical variable indicating whether the contig with the ARG gene is a plasmid.
    + Platon: a logical variable indicating whether the contig with the ARG gene is a mobile genetic element (MGE).

In the different analyses, I aggregated the ARG contig coverage abundances in various ways, using the ARG gene data and the bacterial genus and family taxonomy obtained through Diamond. For alpha and beta diversity analyses, I used the ARG coverage values specific to each gene, while the aggregated values at the genus and family levels were used to establish the overall composition. To aggregate the ARG contigs, I used the following formula:
\[
\text{Aggregated Coverage} = \frac{\sum (\text{Contig Coverage} \times \text{Contig Length})}{\sum (\text{Contig Length})}
\]
Instead for the count-based abundances derived from reads, I aggregated them by simply summing the counts. 

## Load All The Datasets and Libraries

```{r}
library(tidyverse, quietly = TRUE)
```

```{r}
meta_sample <- readRDS("../data/R/ITALY-ILLUMINA/meta_sample.rds") %>% as_tibble(rownames = "sample_id")
abun_bac_sp_reads <- readRDS("../data/R/ITALY-ILLUMINA/abun_bac_sp.rds")
taxa_bac_sp_reads <- readRDS("../data/R/ITALY-ILLUMINA/taxa_bac_sp.rds")
abun_arg_reads_gene <- readRDS("../data/R/ITALY-ILLUMINA/abun_arg_reads_gene.rds")
abun_arg_contigs_gene <- readRDS("../data/R/ITALY-ILLUMINA/abun_arg_contigs_gene.rds")
abun_arg_contigs_genus <- readRDS("../data/R/ITALY-ILLUMINA/abun_arg_contigs_genus.rds")
abun_arg_contigs_family <- readRDS("../data/R/ITALY-ILLUMINA/abun_arg_contigs_family.rds")
```

## Alpha Diversity

```{r}
calculate_pielou <- function(X){
  apply(X, 1, \(x) vegan::diversity(x) / log(length(x[x>0])))
}

df_alpha <- bind_rows(
  
  tibble(source = "bacteria_reads_sp", sample_id = rownames(abun_bac_sp_reads), Pielou = calculate_pielou(abun_bac_sp_reads)),
  tibble(source = "ARG_reads_gene", sample_id = rownames(abun_arg_reads_gene), Pielou = calculate_pielou(abun_arg_reads_gene)),
  tibble(source = "ARG_contigs_gene", sample_id = rownames(abun_arg_contigs_gene), Pielou = calculate_pielou(abun_arg_contigs_gene)),
  
) %>%
  left_join(meta_sample, by = "sample_id")
```

```{r, fig.width=10, fig.height=4}
p_alpha <- df_alpha %>%
  ggplot(aes(x = Surface, y = Pielou, fill = source)) +
  geom_boxplot() +
  facet_wrap(~Surface, scales = "free_x", ncol = 5) +
  theme_bw() +
  theme(legend.position = "top", axis.title.x = element_blank(),
        axis.text.x = element_blank())

p_alpha  
```

```{r}
png(filename = "../plots/diversity_alpha.png", width = 3000, height = 1200, res = 300)
p_alpha
dev.off()

pdf(file = "../plots/diversity_alpha.pdf", width = 10, height = 4)
p_alpha
dev.off()
```

## Beta Diversity - TSNE

```{r}
make_tsne <- function(abun, meta, seed = 42){
  
  set.seed(seed)
  
  p <- abun %>%
    zCompositions::cmultRepl(method = "CZM", z.delete = FALSE) %>%
    vegan::decostand("clr") %>%
    dist %>%
    Rtsne::Rtsne(is_distance=TRUE,
                 perplexity = 4, theta = .25, 
                 pca = TRUE, max_iter = 10000) %>%
    .$Y %>% as_tibble() %>%
    cbind(tibble(sample_id = rownames(abun))) %>%
    relocate(sample_id) %>%
  left_join(meta, by = "sample_id") %>%
  ggplot(aes(x = V1, y = V2, fill = Surface)) +
  geom_point(size = 3, shape = 21) +
  stat_ellipse(aes(group = Surface_Type, color = Surface_Type)) +
  theme_bw() +
  theme(axis.text = element_blank(),
        axis.title = element_blank())
  
  return(p)
}
```

```{r, warning=FALSE, warning=FALSE}
p_beta_tsne_bac_sp_reads <- make_tsne(abun_bac_sp_reads, meta_sample)
p_beta_tsne_arg_reads_gene <- make_tsne(abun_arg_reads_gene, meta_sample)
p_beta_tsne_arg_contigs_gene <- make_tsne(abun_arg_contigs_gene, meta_sample)
```

```{r, fig.width=12, fig.height=4.5}
ggpubr::ggarrange(
  
  p_beta_tsne_bac_sp_reads, p_beta_tsne_arg_reads_gene, p_beta_tsne_arg_contigs_gene,
  labels = c("bacteria species reads", "arg gene reads", "arg gene contigs"),
  common.legend = TRUE, ncol = 3, legend = "top"
  
)
```

```{r}
# PNG
png(filename = "../plots/diversity_beta_tsne_bacteria_reads.png", width = 1500, height = 1350, res = 300)
p_beta_tsne_bac_sp_reads
dev.off()

png(filename = "../plots/diversity_beta_tsne_contigs_reads_gene.png", width = 1500, height = 1350, res = 300)
p_beta_tsne_arg_reads_gene
dev.off()

png(filename = "../plots/diversity_beta_tsne_contigs_contigs_gene.png", width = 1500, height = 1350, res = 300)
p_beta_tsne_arg_contigs_gene
dev.off()

# PDF
pdf(file = "../plots/diversity_beta_tsne_bacteria_reads.pdf", width = 5, height = 4.5)
p_beta_tsne_bac_sp_reads
dev.off()

pdf(file = "../plots/diversity_beta_tsne_arg_reads_gene.pdf", width = 5, height = 4.5)
p_beta_tsne_arg_reads_gene
dev.off()

pdf(file = "../plots/diversity_beta_tsne_arg_contigs_gene.pdf", width = 5, height = 4.5)
p_beta_tsne_arg_contigs_gene
dev.off()
```

## Genera Taxonomy Composition

```{r}
df_composition_bac_sp_reads <- as_tibble(abun_bac_sp_reads, rownames = "sample_id") %>%
  pivot_longer(-sample_id, names_to = "taxa_id", values_to = "abun") %>%
  left_join(as_tibble(taxa_bac_sp_reads, rownames = "taxa_id"), by = "taxa_id") %>%
  group_by(sample_id, genus) %>%
  summarise(rela = sum(abun), .groups = "drop") %>%
  group_by(sample_id) %>%
  mutate(rela = rela/sum(rela)) %>%
  ungroup() %>%
  group_by(genus) %>%
  mutate(top_genus = any(max(rela) >= .05)) %>%
  mutate(Genus = if_else(top_genus, genus, "Other")) %>%
  left_join(meta_sample, by = "sample_id")

df_composition_arg_contigs <- as_tibble(abun_arg_contigs_genus, rownames = "sample_id") %>%
  pivot_longer(-sample_id, names_to = "genus", values_to = "rela") %>%
  group_by(sample_id) %>%
  mutate(rela = rela/sum(rela)) %>%
  ungroup() %>%
  group_by(genus) %>%
  mutate(top_genus = any(max(rela) >= .05)) %>%
  mutate(Genus = if_else(top_genus, genus, "Other")) %>%
  left_join(meta_sample, by = "sample_id") %>%
  mutate(Genus = if_else(Genus == "Unkown", "Other", Genus))

wanted_genus <- c(df_composition_bac_sp_reads$Genus, df_composition_arg_contigs$Genus) %>%
  unique()
wanted_genus <- wanted_genus[-which(wanted_genus=="Other")]
wanted_genus <- c(wanted_genus, "Other")

palette_colors <- rownames(qualpalr::qualpal(n = length(wanted_genus), 
                                             colorspace = list(h = c(0, 360), 
                                                               s = c(.1, .53), 
                                                               l = c(.52,0.85)))$RGB)
names(palette_colors) <- wanted_genus
```


```{r}
df_composition_bac_sp_reads <- df_composition_bac_sp_reads %>%
  mutate(color_Genus = palette_colors[Genus])

df_composition_arg_contigs <- df_composition_arg_contigs %>%
  mutate(color_Genus = palette_colors[Genus])
```

```{r}
p_composition_bac_sp_reads <- df_composition_bac_sp_reads %>%
  ggplot(aes(x = sample_id, y = rela, fill = Genus)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ Surface, scales = "free_x", ncol = 5) +
  scale_fill_manual(values = palette_colors, breaks = names(palette_colors)) +
  theme_bw() +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(),
        axis.title.x = element_blank()) +
  guides(fill=guide_legend(nrow=4))
p_composition_bac_sp_reads
```

```{r}
p_composition_arg_contigs <- df_composition_arg_contigs %>%
  ggplot(aes(x = sample_id, y = rela, fill = Genus)) +
  geom_bar(stat = "identity") +
  facet_wrap(~ Surface, scales = "free_x", ncol = 5) +
  scale_fill_manual(values = palette_colors, breaks = names(palette_colors)) +
  theme_bw() +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, hjust = 1),
        axis.title.y = element_blank(),
        axis.title.x = element_blank()) +
  guides(fill=guide_legend(nrow=4))
p_composition_arg_contigs
```

```{r}
png(filename = "../plots/composition_bacteria_genus_level.png", width = 3000, height = 1500, res = 300)
p_composition_bac_sp_reads
dev.off()

png(filename = "../plots/composition_arg_genus_level.png", width = 3000, height = 1500, res = 300)
p_composition_arg_contigs
dev.off()

pdf(file = "../plots/composition_bacteria_genus_level.pdf", width = 10, height = 5)
p_composition_bac_sp_reads
dev.off()

pdf(file = "../plots/composition_arg_genus_level.pdf", width = 10, height = 5)
p_composition_arg_contigs
dev.off()
```

```{r}
df_arg_contigs_all <- readRDS("../data/R/ITALY-ILLUMINA/arg_contigs_all.rds")
```

```{r}
p_plasmid_waafle <- df_arg_contigs_all %>%
  group_by(sample_id, ARG) %>%
  mutate(is_plasmid = if_else(plasmid_waafle == "LGT", TRUE, FALSE)) %>%
  ungroup() %>%
  select(sample_id, ARG, is_plasmid) %>%
  pivot_wider(names_from = "ARG", values_from = "is_plasmid", values_fill = FALSE, values_fn = any) %>%
  pivot_longer(-sample_id, names_to = "ARG", values_to = "is_plasmid") %>%
  group_by(ARG) %>%
  filter(any(is_plasmid)) %>%
  ungroup() %>%
  left_join(meta_sample, by = "sample_id") %>%
  ggplot() +
  geom_tile(aes(y = sample_id, x = ARG, fill = is_plasmid), color = "gray") +
  scale_fill_viridis_d(name = "LGT", option = "E") +
  ggnewscale::new_scale_fill() +
  geom_tile(aes(y = sample_id, x = 0.4, fill = Surface), width = 0.15) +
  theme_minimal() +
  theme(legend.position = "right",
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        axis.title = element_blank())
p_plasmid_waafle
```

```{r}
p_integron_platon <- df_arg_contigs_all %>%
  group_by(sample_id, ARG) %>%
  mutate(is_mobile = if_else(integron_platon == "MGE", TRUE, FALSE)) %>%
  ungroup() %>%
  select(sample_id, ARG, is_mobile) %>%
  pivot_wider(names_from = "ARG", values_from = "is_mobile", values_fill = FALSE, values_fn = any) %>%
  pivot_longer(-sample_id, names_to = "ARG", values_to = "is_mobile") %>%
  group_by(ARG) %>%
  filter(any(is_mobile)) %>%
  ungroup() %>%
  left_join(meta_sample, by = "sample_id") %>%
  ggplot() +
  geom_tile(aes(y = sample_id, x = ARG, fill = is_mobile), color = "gray") +
  scale_fill_viridis_d(name = "MGE", option = "E") +
  ggnewscale::new_scale_fill() +
  geom_tile(aes(y = sample_id, x = 0, fill = Surface), width = .75) +
  theme_minimal() +
  theme(legend.position = "top",
        axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1),
        axis.title = element_blank())
p_integron_platon
```

```{r}
png(filename = "../plots/heatmap_plasmid_waafle.png", width = 2100, height = 1500, res = 300)
p_plasmid_waafle
dev.off()

pdf(file = "../plots/heatmap_plasmid_waafle.pdf", width = 7, height = 5)
p_plasmid_waafle
dev.off()

png(filename = "../plots/heatmap_integron_platon.png", width = 2700, height = 1500, res = 300)
p_integron_platon
dev.off()

pdf(file = "../plots/heatmap_integron_platon.pdf", width = 9, height = 5)
p_integron_platon
dev.off()
```

