---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
library(ggpubr)
```

```{r}
meta_sample <- readRDS("../data/R/meta_sample.rds")
bac_reads <- readRDS("../data/R/bac_reads.rds") 
arg_reads <- readRDS("../data/R/arg_reads.rds")
arg_contigs <- readRDS("../data/R/arg_contigs.rds")
```


## Alpha Diversity

```{r, fig.width=15, fig.height=5}
# bacteria species reads
#------------------------------------------------------------------------------#
abun_bac_sp_reads <- bac_reads %>%
  select(sample_id, species, abun) %>%
  pivot_wider(names_from = "species", values_from = "abun",values_fill = 0) %>%
  column_to_rownames("sample_id")

df_alpha_1 <- tibble(
  sample_id = rownames(abun_bac_sp_reads),
  Shannon = vegan::diversity(abun_bac_sp_reads),
  Simpson = vegan::diversity(abun_bac_sp_reads, "simpson")
) %>%
  left_join(meta_sample, by = "sample_id")

stat.test.1 <- compare_means(Shannon ~ Surface, data = df_alpha_1,
                             method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = max(df_alpha_1$Shannon) + .25,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

p1 <- df_alpha_1 %>%
  ggplot(aes(y=Shannon, x=Surface, colour=Surface))+
  geom_boxplot(outlier.shape = NA, size = 1.0, alpha = .01) +
  geom_jitter(aes(fill = Surface),  width = 0.2, height=0, pch=21, size = 2.5) +
  stat_pvalue_manual(stat.test.1, tip.length=0.01, bracket.size=0.4,
                     step.increase=c(0.05), label= "{p.label}") +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 14)) +
  scale_y_continuous(breaks = c(2.5, 4.5, 6.5))

# arg reads
#------------------------------------------------------------------------------#
abun_arg_reads_gene <- arg_reads %>%
  select(sample_id, Gene, abun) %>%
  pivot_wider(names_from = "Gene", values_from = "abun",values_fill = 0) %>%
  column_to_rownames("sample_id")

df_alpha_2 <- tibble(
  sample_id = rownames(abun_arg_reads_gene),
  Shannon = vegan::diversity(abun_arg_reads_gene),
  Simpson = vegan::diversity(abun_arg_reads_gene, "simpson")
) %>%
  left_join(meta_sample, by = "sample_id")

stat.test.2 <- compare_means(Shannon ~ Surface, data = df_alpha_2,
                                 method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = max(df_alpha_2$Shannon) + .1,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

p2 <- df_alpha_2 %>%
  ggplot(aes(y=Shannon, x=Surface, colour=Surface))+
  geom_boxplot(outlier.shape = NA, size = 1.0, alpha = .01) +
  geom_jitter(aes(fill = Surface),  width = 0.2, height=0, pch=21, size = 2.5) +
  stat_pvalue_manual(stat.test.2, tip.length=0.01, bracket.size=0.4,
                     step.increase=c(0.05), label= "{p.label}") +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 14))

# arg contigs
#------------------------------------------------------------------------------#
abun_arg_contigs_gene <- arg_contigs %>%
  select(sample_id, Gene, abun) %>%
  group_by(sample_id, Gene) %>%
  summarise(abun = sum(abun), .groups = "drop") %>%
  pivot_wider(names_from = "Gene", values_from = "abun", values_fill = 0) %>%
  column_to_rownames("sample_id")

df_alpha_3 <- tibble(
  sample_id = rownames(abun_arg_contigs_gene),
  Shannon = vegan::diversity(abun_arg_contigs_gene),
  Simpson = vegan::diversity(abun_arg_contigs_gene, "simpson")
) %>%
  left_join(meta_sample, by = "sample_id")

stat.test.3 <- compare_means(Shannon ~ Surface, data = df_alpha_3,
                                 method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = max(df_alpha_3$Shannon) + .15,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

p3 <- df_alpha_3 %>%
  ggplot(aes(y=Shannon, x=Surface, colour=Surface))+
  geom_boxplot(outlier.shape = NA, size = 1.0, alpha = .01) +
  geom_jitter(aes(fill = Surface),  width = 0.2, height=0, pch=21, size = 2.5) +
  stat_pvalue_manual(stat.test.3, tip.length=0.01, bracket.size=0.4,
                     step.increase=c(0.06), label= "{p.label}") +
  theme_bw() +
  theme(axis.title = element_blank(),
        axis.text.x = element_text(size = 12, angle = 45, hjust = 1, vjust = 1),
        axis.text.y = element_text(size = 14)) +
  scale_y_continuous(breaks = c(2, 3, 4))

p_alpha <- ggpubr::ggarrange(p1, p2, p3, nrow = 1, legend = "none", 
                             labels = c("a", "b", "c"), font.label = list(size = 18))
# Annotate the arranged figure with a common y-axis label
p_alpha <- annotate_figure(p_alpha,
                           left = text_grob("Shannon Index", rot = 90, vjust = 1, size = 14))
p_alpha
```


```{r}
png("../plots/plot_alpha.png", width = 4500, height = 1500, res = 300)
p_alpha
dev.off()

pdf("../plots/plot_alpha.pdf", width = 15, height = 5)
p_alpha
dev.off()

bind_rows(
  cbind(df_alpha_1, "Source" = "bacteria", "Level" = "reads", "Type" = "species"),
  cbind(df_alpha_2, "Source" = "amr", "Level" = "reads", "Type" = "genes"),
  cbind(df_alpha_3, "Source" = "amr", "Level" = "contigs", "Type" = "genes")
) %>%
  write_tsv("../tables/diverisity_alpha.tsv")
```