---
title: "R Notebook"
output: html_notebook
---

```{r, message=FALSE}
library(tidyverse)
```

```{r}
arg_reads <- readRDS("../data/R/arg_reads.rds")
arg_contigs <- readRDS("../data/R/arg_contigs.rds")
meta_sample <- readRDS("../data/R/meta_sample.rds")
```

```{r}
sample_cpm_arg_reads <- arg_reads %>%
  group_by(sample_id) %>%
  summarise(sample_cpm = sum(abun), .groups = "drop") %>%
  left_join(meta_sample, "sample_id")

stat.test.reads <- compare_means(sample_cpm ~ Surface_Type, 
                                 data = sample_cpm_arg_reads,
                                 method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = 600,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

stat.test.reads.surface <- compare_means(sample_cpm ~ Surface, 
                                 data = sample_cpm_arg_reads,
                                 method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = 600,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

sample_cpm_arg_contigs <- arg_contigs %>%
  mutate(abun = abun*10^6) %>%
  left_join(meta_sample, "sample_id") %>%
  group_by(sample_id) %>%
  mutate(abun = abun / Reads_count) %>%
  summarise(sample_cpm = sum(abun)) %>%
  left_join(meta_sample, "sample_id")

stat.test.contigs <- compare_means(sample_cpm ~ Surface_Type, 
                                   data = sample_cpm_arg_contigs,
                                   method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = 60,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))

stat.test.contigs.surface <- compare_means(sample_cpm ~ Surface, 
                                 data = sample_cpm_arg_contigs,
                                 method = "wilcox.test", p.adjust.method = "none") %>%
  filter(p <= .1) %>%
  mutate(y.position = 60,
         p.label = case_when(
           p < .01 ~ "<0.01**",
           p < .05 ~ "<0.05*",
           p <= .1 ~ "<0.1"))
```


```{r, fig.width=10, fig.height=5}
p <- ggpubr::ggarrange(
  
  sample_cpm_arg_reads %>%
    ggplot() +
    geom_boxplot(aes(x = Surface_Type, y = sample_cpm, fill = Surface_Type, color = Surface_Type),
                 alpha = .25, outliers = F, color = rgb(0,0,0,.5)) +
    stat_pvalue_manual(stat.test.reads, tip.length=0.01, bracket.size=0.4,
                    step.increase=c(0.06), label= "{p.label}") +
    geom_jitter(aes(x = Surface_Type, y = sample_cpm, fill = Surface_Type, color = Surface_Type),
                width = .2, size = 2) +
    theme_bw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    scale_color_viridis_d(option = "H") +
    scale_fill_viridis_d(option = "H") +
    ylab("AMR gene abundance (CPM)"),
  
  sample_cpm_arg_contigs %>%
    ggplot() +
    geom_boxplot(aes(x = Surface_Type, y = sample_cpm, fill = Surface_Type, color = Surface_Type),
                 alpha = .25, outliers = F, color = rgb(0,0,0,.5)) +
    stat_pvalue_manual(stat.test.contigs, tip.length=0.01, bracket.size=0.4,
                    step.increase=c(0.06), label= "{p.label}") +
    geom_jitter(aes(x = Surface_Type, y = sample_cpm, fill = Surface_Type, color = Surface_Type),
                width = .2, size = 2) +
    theme_bw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    scale_color_viridis_d(option = "H") +
    scale_fill_viridis_d(option = "H") +
    ylab("Coverage of AMR contigs (CPM)"),
  
  common.legend = TRUE, legend = "none",
  labels = c("a", "b")
)
p
```


```{r}
png("../plots/sample_amr_CPM_SurfaceType.png", width = 3000, height = 1500, res = 300)
p
dev.off()

pdf("../plots/sample_amr_CPM_SurfaceType.pdf", width = 10, height = 5)
p
dev.off()
```



```{r, fig.width=10, fig.height=5}
p2 <- ggpubr::ggarrange(
  
  sample_cpm_arg_reads %>%
    ggplot() +
    geom_boxplot(aes(x = Surface, y = sample_cpm, fill = Surface, color = Surface),
                 alpha = .25, outliers = F, color = rgb(0,0,0,.5)) +
    stat_pvalue_manual(stat.test.reads.surface, tip.length=0.01, bracket.size=0.4,
                    step.increase=c(0.06), label= "{p.label}") +
    geom_jitter(aes(x = Surface, y = sample_cpm, fill = Surface, color = Surface),
                width = .2, size = 2) +
    theme_bw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    ylab("AMR gene abundance (CPM)"),
  
  sample_cpm_arg_contigs %>%
    ggplot() +
    geom_boxplot(aes(x = Surface, y = sample_cpm, fill = Surface, color = Surface),
                 alpha = .25, outliers = F, color = rgb(0,0,0,.5)) +
    stat_pvalue_manual(stat.test.contigs.surface, tip.length=0.01, bracket.size=0.4,
                    step.increase=c(0.06), label= "{p.label}") +
    geom_jitter(aes(x = Surface, y = sample_cpm, fill = Surface, color = Surface),
                width = .2, size = 2) +
    theme_bw() +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    ylab("Coverage of AMR contigs (CPM)"),
  
  common.legend = TRUE, legend = "none",
  labels = c("a", "b")
)
p2
```



```{r}
png("../plots/sample_amr_CPM_Surface.png", width = 3000, height = 1500, res = 300)
p2
dev.off()

pdf("../plots/sample_amr_CPM_Surface.pdf", width = 10, height = 5)
p2
dev.off()
```