---
title: "R Notebook"
output: html_notebook
---

```{r}
library(tidyverse)
```

## META SAMPLES

```{r}
meta_sample <- read_tsv("RAW/metadata_all2.txt", show_col_types = FALSE) %>%
  rename(Technology = Tecnology) %>%
  filter(Country == "Italy", Technology == "Illumina") %>%
  select(-all_of(c("Reads_count", "Host", "Level"))) %>%
  relocate(sample_id, .before = 1) %>%
  column_to_rownames("sample_id") %>%
  mutate(Surface_Type = case_when(
    str_detect(Surface, "clean") ~ "clean",
    str_detect(Surface, "dirty") ~ "dirty", 
    str_detect(Surface, "carcass") ~ "carcass"
  )) %>%
  rownames_to_column("sample_id") %>%
  mutate(Surface = factor(Surface, levels = c("dirty drain", 
                                              "dirty surface",
                                              "clean drain",
                                              "clean surface", 
                                              "carcass")))

saveRDS(meta_sample, "R/meta_sample.rds")
```

## ARG READS

```{r}
df_arg_reads <- read_tsv("RAW/ARGs_cpm_ITA_ILL.txt", show_col_types = FALSE) %>%
  pivot_longer(-c(Antibiotic_Family, Gene), names_to = "sample_id", values_to = "abun") %>%
  filter(sample_id %in% meta_sample$sample_id) %>%
  filter(abun > 0) %>%
  select(sample_id, Gene, abun, Antibiotic_Family) %>%
  left_join(meta_sample, by = "sample_id")

# Used only for check
# df_arg_reads_raw %>%
#   group_by(sample_id, Gene) %>%
#   summarise(n = n()) %>%
#   arrange(desc(n))
# df_arg_reads_raw %>%
#   select(-Antibiotic_Family) %>%
#   pivot_wider(names_from = "Gene", values_from = "abun", values_fill = 0) %>%
#   View
```

## ARG CONTIGS 

```{r}
df_arg_contigs <- read_tsv("RAW/data_merged_contigs_final.txt", show_col_types = FALSE) %>%
  select(-1) %>%
  filter(Country == "Italy", Technology == "illumina") %>%
  rename(sample_id = Sample) %>%
  filter(sample_id %in% meta_sample$sample_id) %>%
  relocate(sample_id, .before = 1) %>%
  mutate(plasmid_waafle = if_else(grepl("-", SYNTENY), "", "LGT"),
         integron_platon = if_else(grepl("-", Platon_ID), "", "MGE")) %>%
  mutate(phylum_diamond = str_extract(Diamond_nr, "p_[^;]+"),
         phylum_diamond = str_remove(phylum_diamond, "p_"),
         phylum_diamond = if_else(is.na(phylum_diamond), "Unknown", phylum_diamond)) %>%
  mutate(family_diamond = str_extract(Diamond_nr, "f_[^;]+"),
         family_diamond = str_remove(family_diamond, "f_"),
         family_diamond = if_else(is.na(family_diamond), "Unknown", family_diamond)) %>%
  mutate(genus_diamond = str_extract(Diamond_nr, "g_[^;]+"),
         genus_diamond = str_remove(genus_diamond, "g_"),
         genus_diamond = if_else(is.na(genus_diamond), "Unknown", genus_diamond)) %>%
  mutate(species_diamond = str_extract(Diamond_nr, "s_[^;]+"),
         species_diamond = str_remove(species_diamond, "s_"),
         species_diamond = if_else(is.na(species_diamond), "Unknown", species_diamond)) %>%
  rename(Antibiotic_Family = Fam) %>%
    mutate(abun = str_extract(Contig_name, "cov_\\d+\\.\\d+"),
           abun = as.numeric(str_replace(abun, "cov_", ""))) %>%
  select(-any_of(c("Country","Time","Surface","Technology","Surface_Type"))) %>%
  left_join(meta_sample, by = "sample_id") %>%
  relocate(sample_id, Contig_name, abun, Surface, ARG, Gene, Antibiotic_Family, 
           species_diamond, genus_diamond, family_diamond, phylum_diamond, 
           plasmid_waafle, integron_platon) %>%
  mutate(across(c(phylum_diamond, genus_diamond, family_diamond), ~if_else(is.na(.), "Unknown", .)))
```

## BACTERIA READS

```{r}
df_bac_species <- read_tsv("RAW/matrix_bacteria_S_idcorretto.txt", show_col_types = FALSE) %>%
  column_to_rownames("...1") %>%
  t %>% as_tibble(rownames = "sample_id") %>%
  column_to_rownames("sample_id") %>%
  rename("Ochrobactrum quorumnocens" = "[Ochrobactrum] quorumnocens",
         "Roseinatronobacter bogoriensis subsp. barguzinensis" = "Rhodobaca barguzinensis")

df_bac_species <- df_bac_species[meta_sample$sample_id, ]
df_bac_species <- df_bac_species[, colSums(df_bac_species) >0]

taxa_bacteria <- taxizedb::name2taxid(colnames(df_bac_species)) %>%
  taxizedb::classification(verbose = T)

# Assuming you already have `taxa_bacteria` as the result of `taxizedb::classification()`
# Define the taxonomic ranks you expect in the final data frame
expected_ranks <- c("superkingdom", "phylum", "class", "order", "family", "genus", "species")

# Store the original names from colnames(df_bac_species)
original_names <- colnames(df_bac_species)

# Function to safely process each element and attach the original name
process_taxa <- function(x, original_name) {
  # Check if the element is a data frame and has the necessary columns
  if (is.data.frame(x) && all(c("rank", "name") %in% colnames(x))) {
    df <- as.data.frame(t(deframe(x[, c("rank", "name")])))
    
    # Ensure all expected ranks are present, fill missing ones with NA
    missing_ranks <- setdiff(expected_ranks, colnames(df))
    if (length(missing_ranks) > 0) {
      df[missing_ranks] <- NA
    }
    
    # Add the original name as a new column
    df$original_name <- original_name
    
    return(df[, c("original_name", expected_ranks), drop = FALSE])  # Return data frame with original name and all expected ranks
  } else {
    # Return a data frame with NA for non-conforming elements, and include the original name
    return(as.data.frame(matrix(NA, nrow = 1, ncol = length(expected_ranks) + 1, 
                                dimnames = list(NULL, c("original_name", expected_ranks)))) %>%
             mutate(original_name = original_name))
  }
}

# Apply the function to each element of the list with corresponding original name
taxa_bacteria <- map2(taxa_bacteria, original_names, process_taxa) %>%
  bind_rows() 
colnames(taxa_bacteria)[8] <- "species_taxizedb"
```


```{r}
df_bac_reads <- df_bac_species %>%
  rownames_to_column("sample_id") %>%
  pivot_longer(-sample_id, names_to = "species", values_to = "abun") %>%
  left_join(meta_sample, by = "sample_id") %>%
  left_join(taxa_bacteria, join_by(species == original_name))

df_bac_reads <- df_bac_reads %>%
  mutate(across(c(phylum, genus, family), ~if_else(is.na(.), "Unknown", .)))
```


```{r}
saveRDS(meta_sample, "R/meta_sample.rds")
saveRDS(df_arg_contigs, "R/arg_contigs.rds")
saveRDS(df_arg_reads, "R/arg_reads.rds")
saveRDS(df_bac_reads, "R/bac_reads.rds")
```



